tests/transports/stream_tests.nim

```diff
@@ -220,25 +205,140 @@ template streamTransportTest*(
+        for i in 0 ..< parallelWrites:
+          let offset = i * messageSize
+          let expectedValue = receivedData[offset]
+          # all elements of same data have to be the same
+          for j in 0 ..< messageSize:
+            check receivedData[offset + j] == expectedValue
+            if receivedData[offset + j] != expectedValue:
+              break # stop on first mismatch (not to pollute stdout)
+
+    proc clientStreamHandler(stream: Connection) {.async: (raises: []).} =
+      noExceptionWithStreamClose(stream):
+        var writeFuts: seq[Future[void]] = @[]
+        for i in 0 ..< parallelWrites:
+          # each write has to have unique data
+          let message = newData(messageSize, uint8(i + 1))
+          let fut = stream.write(message)
+          writeFuts.add(fut)
+        await allFutures(writeFuts)
 
     await runSingleStreamScenario(
       ma, transportProvider, streamProvider, serverStreamHandler, clientStreamHandler
     )
+
+  asyncTest "server with multiple parallel connections":
+    let ma = @[MultiAddress.init(address).tryGet()]
+    const numConnections = 5
+
+    # Track when stream handlers complete
+    var serverStreamHandlerFuts: array[numConnections, Future[void]]
+    for i in 0 ..< numConnections:
+      serverStreamHandlerFuts[i] = newFuture[void]()
+
+    proc serverHandler(server: Transport) {.async.} =
+      # Accept multiple connections and handle them
+      var futs: seq[Future[void]]
+      for i in 0 ..< numConnections:
+        let conn = await server.accept()
+        let muxer = streamProvider(server, conn)
+
+        # Use a proc to properly capture loop index
+        proc setupConnection(conn: Connection, muxer: Muxer, handlerIndex: int) =
+          muxer.streamHandler = proc(stream: Connection) {.async: (raises: []).} =
+            noExceptionWithStreamClose(stream):
+              var buffer: array[1, byte]
+              await stream.readExactly(addr buffer, 1)
+              await stream.write(@buffer)
+
+              # Signal that this stream handler is done
+              serverStreamHandlerFuts[handlerIndex].complete()
+
+          let startStreamHandlerAndCleanup = proc() {.async.} =
+            let muxerTask = muxer.handle()
+
+            await muxerTask
+
+            # Wait for the stream handler to complete before closing
+            await serverStreamHandlerFuts[handlerIndex]
+
+            await muxer.close()
+            await conn.close()
+
+          futs.add(startStreamHandlerAndCleanup())
+
+        setupConnection(conn, muxer, i)
+      await allFutures(futs)
+
+    proc runClient(server: Transport, connectionId: int) {.async.} =
+      let client = transportProvider()
+      let conn = await client.dial(server.addrs[0])
+      let muxer = streamProvider(client, conn)
+      let muxerTask = muxer.handle()
+
+      let stream = await muxer.newStream()
+      await stream.write(@[byte(connectionId)])
```

Comment:
note, comment from above can be inverted. that is server can read data without delay. and delay can be introduced on client side:

```nim
for i in chunksCount:
 await stream.write(newData(i))
 away asyncSleep(100.milliseconds)
```

the delay on client side should introduce blocking at server when chunk is read (server waits for next chunk), this will cause that future to get blocked which will give control to other future.